closed/src/java.base/share/classes/sun/security/rsa/NativeRSACore.java

```diff
@@ -138,17 +140,40 @@ public static int getByteLength(BigInteger b) {
             outputLen = getByteLength(n);
             output = new byte[outputLen];
             int verifyInt = verify ? outputLen : -1;
-            outputLen = nativeCrypto.RSADP(msg, msg.length, output, verifyInt, nativePtr);
+            RSADPBigNumRPointer rsadpBn = rsadpBnPtr.get();
+            if (rsadpBn == null) {
+                rsadpBn = new RSADPBigNumRPointer();
+                rsadpBnPtr.set(rsadpBn);
+            }
+            outputLen = nativeCrypto.RSADP(msg, msg.length, output, verifyInt, nativePtr, rsadpBn.getBN());
         } finally {
             key.returnNativePtr(nativePtr);
         }
-
         if (outputLen == -1) {
-            return null;
-        } else if (outputLen == -2) {
             throw new BadPaddingException("RSA private key operation failed");
         }
 
         return output;
     }
+
+    static class RSADPBigNumRPointer {
+        long rsadpBN = 0;
+
+        RSADPBigNumRPointer() {
+            this.rsadpBN = nativeCrypto.create_rsadp_BN();
```

Comment:
Using `finalize` is deprecated. Please use a `Cleaner`, like in `NativeDigest.java`.

Reply:
We should also consider wrapping the RSADPBigNumRPointer in a SoftReference, so they don't stick around forever on all the threads. I see jdk.internal.ref.CleanerImpl.SoftCleanableRef() can be used for this.

Reply:
@pshipton I added this soft reference in my most recent commit. Based on your suggestion, I also created [this test](https://gist.github.ibm.com/samasri/833b75b77affaa474e27c683f30c6ed7) to make sure the code is working as expected.

The test works as follows:
1. Calls the `NativeRSACore.sign()` function
2. Calls `System.gc()` 32 times (which is supposed to trigger the GC to work, this should effectively delete the soft reference from memory)
3. Calls `NativeRSACore.sign()` again

I also added a print statement in the JDK at the time of the creation of the `RSADPBigNumRPointer` (specifically, I added it below [this line](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/89799461d95cf16481cabed5fc1e6dba3f2689ed/closed/src/java.base/share/classes/sun/security/rsa/NativeRSACore.java#L150) inside the if statement). When running the test, I can see that this statement is printed twice, which proves that it does get deleted by the GC in the right circumstances. Furthermore, I tried removing the `System.gc()` for loop from the test, and the print statement that I added in the JDK only appeared once. This is additional proof that the soft pointer is created twice because of the calls of `System.gc()`, which triggers the garbage collector.

Reply:
Thanks, lgtm. If you haven't already, it's worth adding a print in the cleaner to ensure it's working to clean up. You'll need the 32 System.gc()'s to trigger the cleaning.

Reply:
Actually, to avoid recompiling the code, you can use a method tracepoint instead of a print to ensure RSADPBigNumRPointerCleanerRunnable.run() is happening.

Reply:
I added a print statement in [this function](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/89799461d95cf16481cabed5fc1e6dba3f2689ed/closed/src/java.base/share/classes/sun/security/rsa/NativeRSACore.java#L197-L199) and I can confirm that I can see the message when I have `System.gc()` in the test. And I don't see the statement when `System.gc` is not in the test.